### PR TITLE
14870 mobile API logout

### PR DESCRIPTION
**Pull request by kreek**
## Description of change
Adds the `/mobile/v0/user/logout` endpoint which revokes the user's IAM SSOe OAuth access token and then clears the IAM user, user identity, and session objects from redis.

## Original issue(s)
department-of-veterans-affairs/va.gov-team#14870

## Things to know about this PR
* Adds rails logs for logout success or failure
* Mobile swagger docs updated

testing done: local unit and integration tests
testing planned: staging integration tests

### CODE REVIEW

File: modules/mobile/app/controllers/mobile/v0/users_controller.rb [hunk 1]
```diff
@@ -9,6 +9,13 @@ def show
         render json: Mobile::V0::UserSerializer.new(@current_user, options)
       end
 
+      def logout
+        session_manager = IAMSSOeOAuth::SessionManager.new(access_token)
+        session_manager.logout
+
+        head(:ok)
```
Comment (jjulian) [inline modules/mobile/app/controllers/mobile/v0/users_controller.rb:16]: 👍 for a nice short response!

File: modules/mobile/spec/support/iam_session_helper.rb [hunk 2]
```diff
@@ -36,6 +36,7 @@ def iam_sign_in(iam_user = FactoryBot.build(:iam_user), access_token = nil)
 
   config.before :each, type: :request do
     Flipper.enable('mobile_api')
+    Flipper.enable('va_online_scheduling')
```
Comment (jjulian) [inline modules/mobile/spec/support/iam_session_helper.rb:39]: Is this supposed to be part of this PR?

File: lib/iam_ssoe_oauth/configuration.rb [hunk 3]
```diff
@@ -35,6 +35,7 @@ def connection
         base_path, headers: base_request_headers, request: request_options, ssl: ssl_options
       ) do |conn|
         conn.use :breakers
+        conn.use Faraday::Response::RaiseError
```
Comment (jjulian) [inline lib/iam_ssoe_oauth/configuration.rb:38]: Does this change impact the introspect call?

File: lib/iam_ssoe_oauth/session_manager.rb [hunk 4]
```diff
@@ -15,8 +15,34 @@ def find_or_create_user
...
 
+    def destroy_user_session
+      uuid = @session.uuid
+
+      identity_destroy_count = IAMUserIdentity.find(uuid).destroy
+      user_destroy_count = IAMUser.find(uuid).destroy
+      session_destroy_count = @session.destroy
+
+      # redis returns number of records successfully deleted
+      if [identity_destroy_count, user_destroy_count, session_destroy_count].sum == 3
```
Comment (jjulian) [inline lib/iam_ssoe_oauth/session_manager.rb:33]: Is it possible one of these counts gets returned as 2? That could throw off the sum and lead to a false positive.

**Review by jjulian — CHANGES_REQUESTED** [review]
Looks good, a couple detailed questions before we move forward

File: lib/iam_ssoe_oauth/configuration.rb (hunk 3 above)
Comment (kreek) [reply lib/iam_ssoe_oauth/configuration.rb:38]: It does, although an inactive response from the introspect call is a 200 and doesn't throw an error (we catch it and return unauthorized). I could add the same 500-> 502 catch there.

File: lib/iam_ssoe_oauth/session_manager.rb (hunk 4 above)
Comment (kreek) [reply lib/iam_ssoe_oauth/session_manager.rb:33]: technically possible 😄 I had it as `.all? { |c| c == 1 } ` before I'll change it to != 0

File: modules/mobile/spec/support/iam_session_helper.rb (hunk 2 above)
Comment (kreek) [reply modules/mobile/spec/support/iam_session_helper.rb:39]: no, removed

File: lib/iam_ssoe_oauth/session_manager.rb [hunk 5]
```diff
@@ -15,8 +15,34 @@ def find_or_create_user
...
 
+    def destroy_user_session
+      uuid = @session.uuid
+
+      identity_destroy_count = IAMUserIdentity.find(uuid).destroy
+      user_destroy_count = IAMUser.find(uuid).destroy
+      session_destroy_count = @session.destroy
+
+      # redis returns number of records successfully deleted
+      if [identity_destroy_count, user_destroy_count, session_destroy_count].all? { |c| c != 0 }
```
Comment (jjulian) [inline lib/iam_ssoe_oauth/session_manager.rb:33]:
```suggestion
      if [identity_destroy_count, user_destroy_count, session_destroy_count].all? { |c| c > 0 }
```
Reads "if all of these are greater than zero" which I think is mildly more readable than "if all of these are not equal to zero"

File: lib/iam_ssoe_oauth/session_manager.rb [hunk 6]
```diff
@@ -15,8 +15,34 @@ def find_or_create_user
...
+        Rails.logger.info('IAMUser log out success', uuid: uuid)
+        true
+      else
+        Rails.logger.warn('IAMUser log out failure', uuid: uuid, status: {
```
Comment (jjulian) [inline lib/iam_ssoe_oauth/session_manager.rb:37]: Should the api consumer make an effort not to call /logout unless they think a session is in-progress? Looks like the only side-effect is this warning log message.

**Review by jjulian — APPROVED** [review]

**Review by thilton-oddball — APPROVED** [review]
I think I followed everything on this and it looks good.

**Pull request merged**